### ISSUE TITLE
Use of NameValidator in np.genfromtxt is inconsistent with the rules for naming structured array fields

### ISSUE THREAD

**Issue by alimuldal** [numpy/numpy#5686]
This issue has cropped up a [couple][1] of [times][2] before on StackOverflow. `np.genfromtxt` automatically modifies field names by replacing spaces and stripping out certain non-alphanumeric characters etc.:

```python
import numpy as np
from io import BytesIO

s = 'name,name with spaces,(x-1)!\n1,2,3\n4,5,6'
x = np.genfromtxt(BytesIO(s), delimiter=',', names=True)
print(repr(x))
# array([(1.0, 2.0, 3.0), (4.0, 5.0, 6.0)], 
#       dtype=[('name', '<f8'), ('name_with_spaces', '<f8'), ('x1', '<f8')])
```

However, these are all perfectly legal field names:

```python
names = ['name', 'name with spaces', '(x-1)!']
types = ('f',) * 3
dtype = zip(names, types)

x2 = np.empty(2, dtype=dtype)
x2[:] = [(1.0, 2.0, 3.0), (4.0, 5.0, 6.0)]
print(repr(x2))
# array([(1.0, 2.0, 3.0), (4.0, 5.0, 6.0)], 
#       dtype=[('name', '<f4'), ('name with spaces', '<f4'), ('(x-1)!', '<f4')])
print(x2['(x-1)!'])
# [3. 6.]
```

The rationale behind mangling field names in this way is not clear to me. In my opinion, it should either be the case that spaces and non-alphanumeric characters are disallowed in field names altogether (and my second example should raise an exception) , or `np.genfromtxt` should leave field names alone by default.

[1]:http://stackoverflow.com/q/29097917/1461210
[2]:http://stackoverflow.com/q/16020137/1461210

**Comment by charris** [comment]
I think this is probably a defect, but I'm not sure ;) Could you send a post to the mailing list for discussion?

**Comment by alimuldal** [comment]
@charris I posted to the mailing list a couple of days ago regarding this issue, but no responses so far. Have you had any more thoughts?

**Comment by swook** [comment]
I encountered the same issue and found that it was resolved in 1b88f08b4f27e26adc7a1200a51974eb5bf719e0

**Comment by alimuldal** [comment]
@swook I don't believe this is resolved at all. That commit is dated about a year before I even opened the issue!

Another question cropped up on SO today: http://stackoverflow.com/q/35234643/1461210. It would be great if we could have some clarification over whether or not this behavior should be considered a defect.

@charris ?

**Comment by njsmith** [comment]
@alimuldal: Are you sure? The tag display underneath that commit suggests that the first release including it was 1.10, which just came out in October. (Though it could have been backported, or maybe you were testing directly against master, I dunno.)

**Comment by alimuldal** [comment]
@njsmith  I'm using 1.10.4, and I also see the same thing in master. Just to be clear, I'm asking:

* Whether `genfromtxt` should be doing any field name mangling *at all*, given that this seems to be inconsistent with the actual rules (whatever those may be...) for naming fields in structured arrays.

* If there is an important reason to mangle field names, can we change the default behaviour to either not clobber user-specified field names, give some sort of warning, or at least make the documentation more explicit?

**Comment by njsmith** [comment]
Oh, well, sorry, in that case I dunno :-). If you're working with heterogenous named data, then for most purposes I'd recommend `pandas` as being a more modern and featureful alternative to structured arrays.
